// caterpillar/out/lib/logger.js
// Generated by CoffeeScript 1.6.2
var Logger, extendr, stream,
  __bind = function(fn, me){ return function(){ return fn.apply(me, arguments); }; },
  __hasProp = {}.hasOwnProperty,
  __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; },
  __slice = [].slice;

extendr = require('extendr');

stream = require('stream');

Logger = (function(_super) {
  __extends(Logger, _super);

  Logger.prototype.config = {
    levels: {
      emergency: 0,
      alert: 1,
      critical: 2,
      error: 3,
      warning: 4,
      notice: 5,
      info: 6,
      debug: 7,
      emerg: 0,
      crit: 2,
      err: 3,
      warn: 4,
      note: 5,
      "default": 6
    }
  };

  function Logger(opts) {
    this.setConfig = __bind(this.setConfig, this);    this.setConfig(opts);
    Logger.__super__.constructor.apply(this, arguments);
  }

  Logger.prototype.pipe = function(child) {
    if (typeof child.setConfig === "function" ? child.setConfig(this.config) : void 0) {
      this.on('config', child.setConfig);
    }
    return Logger.__super__.pipe.apply(this, arguments);
  };

  Logger.prototype.setConfig = function(opts) {
    this.config = extendr.deepExtend({}, this.config, opts);
    return this;
  };

  Logger.prototype.getConfig = function() {
    return this.config;
  };

  Logger.prototype._transform = function(chunk, encoding, next) {
    return next(null, chunk);
  };

  Logger.prototype.getLevelCode = function(name) {
    var _ref;

    return (_ref = this.config.levels[name]) != null ? _ref : null;
  };

  Logger.prototype.getLevelName = function(code) {
    var key, value, _ref;

    _ref = this.config.levels;
    for (key in _ref) {
      if (!__hasProp.call(_ref, key)) continue;
      value = _ref[key];
      if (value === code) {
        return key;
      }
    }
    return null;
  };

  Logger.prototype.getLevelInfo = function(level) {
    var levelCode, levelName, result;

    result = {
      levelCode: null,
      levelName: null
    };
    if (typeof level === 'number') {
      levelCode = level;
      levelName = this.getLevelName(levelCode);
    } else {
      levelName = level;
      levelCode = this.getLevelCode(levelName);
      levelName = this.getLevelName(levelCode);
      if (levelCode == null) {
        levelCode = this.getLevelCode('default');
        levelName = this.getLevelName(levelCode);
      }
    }
    result.levelCode = levelCode;
    result.levelName = levelName;
    return result;
  };

  Logger.prototype.getLineInfo = function() {
    var err, line, lines, parts, result, _i, _len;

    result = {
      line: -1,
      method: 'unknown',
      file: 'unknown'
    };
    err = new Error();
    lines = err.stack.split('\n');
    for (_i = 0, _len = lines.length; _i < _len; _i++) {
      line = lines[_i];
      if (line.indexOf(__dirname) !== -1 || line.indexOf(' at ') === -1) {
        continue;
      }
      parts = line.split(':');
      if (parts[0].indexOf('(') === -1) {
        result.method = 'unknown';
        result.file = parts[0].replace(/^.+?\s+at\s+/, '');
      } else {
        result.method = parts[0].replace(/^.+?\s+at\s+/, '').replace(/\s+\(.+$/, '');
        result.file = parts[0].replace(/^.+?\(/, '');
      }
      result.line = parts[1];
      break;
    }
    return result;
  };

  Logger.prototype.format = function() {
    var args, entry, key, level, value, _ref, _ref1;

    level = arguments[0], args = 2 <= arguments.length ? __slice.call(arguments, 1) : [];
    entry = {};
    entry.args = args;
    entry.date = new Date().toISOString();
    _ref = this.getLevelInfo(level);
    for (key in _ref) {
      if (!__hasProp.call(_ref, key)) continue;
      value = _ref[key];
      entry[key] = value;
    }
    _ref1 = this.getLineInfo(level);
    for (key in _ref1) {
      if (!__hasProp.call(_ref1, key)) continue;
      value = _ref1[key];
      entry[key] = value;
    }
    return entry;
  };

  Logger.prototype.log = function() {
    var args, entry, level;

    level = arguments[0], args = 2 <= arguments.length ? __slice.call(arguments, 1) : [];
    entry = this.format.apply(this, [level].concat(__slice.call(args)));
    this.write(JSON.stringify(entry));
    return this;
  };

  return Logger;

})(stream.Transform);

module.exports = {
  Logger: Logger
};
