// docpad-plugin-paged/out/paged.plugin.js
// Generated by CoffeeScript 1.6.3
var __hasProp = {}.hasOwnProperty,
  __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; };

module.exports = function(BasePlugin) {
  var PagedPlugin, TaskGroup, _ref;
  TaskGroup = require('taskgroup').TaskGroup;
  return PagedPlugin = (function(_super) {
    __extends(PagedPlugin, _super);

    function PagedPlugin() {
      _ref = PagedPlugin.__super__.constructor.apply(this, arguments);
      return _ref;
    }

    PagedPlugin.prototype.name = 'paged';

    PagedPlugin.prototype.extendCollections = function(opts) {
      var docpad, me;
      me = this;
      docpad = this.docpad;
      docpad.getFiles({
        isPaged: true,
        isPagedAuto: {
          $ne: true
        }
      }).on('remove', function(model) {
        return me.removePagesFor(model);
      });
      return this;
    };

    PagedPlugin.prototype.extendTemplateData = function(opts) {
      var docpad, templateData;
      docpad = this.docpad;
      templateData = opts.templateData;
      templateData.getPageUrl = function(pageNumber, document) {
        var err, page, pageDocument, pageId, pageUrl, relativePath, _ref1;
        if (document == null) {
          document = this.getDocument();
        }
        page = document.get('page');
        if (pageNumber == null) {
          pageNumber = (_ref1 = page != null ? page.number : void 0) != null ? _ref1 : 0;
        }
        pageId = page.pages[pageNumber];
        pageDocument = docpad.getFileById(pageId);
        if (pageDocument == null) {
          relativePath = document.get('relativePath');
          err = "Could not find document with id " + pageId + " that is page " + pageNumber + " of " + relativePath;
          docpad.error(err);
          pageUrl = err;
        } else {
          pageUrl = pageDocument.get('url');
        }
        return pageUrl;
      };
      templateData.hasNextPage = function(document) {
        var has, page;
        if (document == null) {
          document = this.getDocument();
        }
        page = document.get('page');
        has = page.number < page.count - 1;
        return has;
      };
      templateData.getNextPage = function(document) {
        var page, result;
        if (document == null) {
          document = this.getDocument();
        }
        page = document.get('page');
        result = '#';
        if (page.number < page.count - 1) {
          result = this.getPageUrl(page.number + 1, document);
        }
        return result;
      };
      templateData.hasPrevPage = function(document) {
        var has, page;
        if (document == null) {
          document = this.getDocument();
        }
        page = document.get('page');
        has = page.number > 0;
        return has;
      };
      templateData.getPrevPage = function(document) {
        var page, result;
        if (document == null) {
          document = this.getDocument();
        }
        page = document.get('page');
        result = '#';
        if (page.number > 0) {
          result = this.getPageUrl(page.number - 1, document);
        }
        return result;
      };
      return true;
    };

    PagedPlugin.prototype.removePagesFor = function(document, collection, next) {
      var database, docpad, pages, tasks, _ref1;
      docpad = this.docpad;
      database = docpad.getDatabase();
      tasks = new TaskGroup();
      if (next) {
        tasks.once('complete', next);
      }
      pages = ((_ref1 = document.get('page')) != null ? _ref1.pages : void 0) || [];
      pages.forEach(function(pageId) {
        var pageDocument;
        if (pageId === document.id) {
          return;
        }
        pageDocument = database.get(pageId);
        if (!pageDocument) {
          return;
        }
        if (collection != null) {
          collection.remove(pageDocument);
        }
        database.remove(pageDocument);
        return tasks.addTask(function(complete) {
          return pageDocument["delete"](complete);
        });
      });
      tasks.run();
      return this;
    };

    PagedPlugin.prototype.renderBeforePriority = 550;

    PagedPlugin.prototype.renderBefore = function(opts, next) {
      var collection, database, docpad, me, newPagesToRender, sourcePageDocuments, tasks, templateData;
      me = this;
      docpad = this.docpad;
      collection = opts.collection, templateData = opts.templateData;
      database = docpad.getDatabase();
      newPagesToRender = [];
      tasks = new TaskGroup().once('complete', next);
      sourcePageDocuments = collection.findAll({
        isPaged: true,
        isPagedAuto: {
          $ne: true
        }
      });
      sourcePageDocuments.forEach(function(document) {
        return tasks.addTask(function(complete) {
          return me.removePagesFor(document, collection, complete);
        });
      });
      sourcePageDocuments.forEach(function(document) {
        return tasks.addGroup(function(addGroup, addTask) {
          var basename, extension, filename, lastDoc, meta, numberOfPages, outBasename, outExtension, outFilename, pageSize, pagedCollection, pagedCollectionName, pages, relativePath, url, _i, _results;
          meta = document.getMeta();
          numberOfPages = meta.get('pageCount') || 1;
          pageSize = meta.get('pageSize') || 1;
          lastDoc = pageSize * numberOfPages;
          if (meta.get('pagedCollection')) {
            pagedCollectionName = meta.get('pagedCollection');
            pagedCollection = docpad.getCollection(pagedCollectionName);
            numberOfPages = Math.ceil(pagedCollection.length / pageSize);
            lastDoc = pagedCollection.length;
          }
          relativePath = document.get('relativePath');
          filename = document.get('filename');
          basename = document.get('basename');
          extension = document.get('extensions').join('.');
          outFilename = document.get('outFilename');
          outBasename = document.get('outBasename');
          outExtension = document.get('outExtension');
          url = document.get('url');
          pages = [document.id];
          document.set({
            isPaged: true,
            isPagedAuto: false,
            isPagedFor: false,
            page: {
              count: numberOfPages,
              size: pageSize,
              number: 0,
              startIdx: 0,
              endIdx: Math.min(pageSize, lastDoc),
              pages: pages
            }
          });
          if (numberOfPages > 1) {
            return (function() {
              _results = [];
              for (var _i = 1; 1 <= numberOfPages ? _i < numberOfPages : _i > numberOfPages; 1 <= numberOfPages ? _i++ : _i--){ _results.push(_i); }
              return _results;
            }).apply(this).forEach(function(pageNumber) {
              return addTask(function(complete) {
                var pageDocument, pageFilename, pageOutFilename, pageRelativePath;
                pageDocument = document.clone();
                pageFilename = "" + basename + "-" + pageNumber + "." + extension;
                pageOutFilename = "" + outBasename + "." + pageNumber + "." + outExtension;
                pageRelativePath = relativePath.replace(filename, pageFilename);
                pageDocument.attributes.urls = [];
                pageDocument.set({
                  isPagedAuto: true,
                  isPagedFor: document.id,
                  page: {
                    count: numberOfPages,
                    size: pageSize,
                    number: pageNumber,
                    startIdx: pageNumber * pageSize,
                    endIdx: Math.min(pageNumber * pageSize + pageSize, lastDoc),
                    pages: pages
                  }
                });
                pageDocument.setMeta({
                  fullPath: null,
                  relativePath: pageRelativePath,
                  filename: pageFilename,
                  outFilename: pageOutFilename
                });
                return pageDocument.normalize(function(err) {
                  if (err) {
                    return complete(err);
                  }
                  pages.push(pageDocument.id);
                  collection.add(pageDocument);
                  database.add(pageDocument);
                  return complete();
                });
              });
            });
          }
        });
      });
      tasks.run();
      return true;
    };

    return PagedPlugin;

  })(BasePlugin);
};
