// joe/out/lib/reporters/list.js
// Generated by CoffeeScript 1.4.0
var ConsoleReporter, ListReporter, isWindows,
  __hasProp = {}.hasOwnProperty,
  __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; };

ConsoleReporter = typeof require !== "undefined" && require !== null ? require(__dirname + '/console') : this.joe.ConsoleReporter;

isWindows = (typeof process !== "undefined" && process !== null) && process.platform.indexOf('win') === 0;

ListReporter = (function(_super) {

  __extends(ListReporter, _super);

  ListReporter.prototype.startSuite = function() {};

  ListReporter.prototype.startTest = function() {};

  ListReporter.prototype.finishSuite = function() {};

  function ListReporter(config) {
    var _base, _base1, _ref, _ref1;
    this.config || (this.config = config || {});
    if ((_ref = (_base = this.config).fail) == null) {
      _base.fail = isWindows ? 'ERR! ' : '✘  ';
    }
    if ((_ref1 = (_base1 = this.config).pass) == null) {
      _base1.pass = isWindows ? 'OK   ' : '✔  ';
    }
    ListReporter.__super__.constructor.apply(this, arguments);
  }

  ListReporter.prototype.finishTest = function(suite, testName, err) {
    var check, message;
    testName = this.getTestName(suite, testName);
    if (!testName) {
      return this;
    }
    check = (err ? this.config.fail : this.config.pass);
    message = "" + check + testName;
    console.log(message, (typeof process !== "undefined" && process !== null) === false && err ? [err, err.stack] : '');
    return this;
  };

  ListReporter.prototype.exit = function(exitCode) {
    var err, errorLog, errorLogs, index, suite, testName, totalErrors, totalFailedTests, totalIncompleteTests, totalPassedTests, totalTests, _i, _len, _ref;
    _ref = this.joe.getTotals(), totalTests = _ref.totalTests, totalPassedTests = _ref.totalPassedTests, totalFailedTests = _ref.totalFailedTests, totalIncompleteTests = _ref.totalIncompleteTests, totalErrors = _ref.totalErrors;
    if (exitCode) {
      errorLogs = this.joe.getErrorLogs();
      console.log("\n" + this.config.summaryFail, totalPassedTests, totalTests, totalFailedTests, totalIncompleteTests, totalErrors);
      for (index = _i = 0, _len = errorLogs.length; _i < _len; index = ++_i) {
        errorLog = errorLogs[index];
        suite = errorLog.suite, testName = errorLog.testName, err = errorLog.err;
        testName = this.getTestName(suite, testName);
        console.log("\n" + this.config.failHeading, index + 1);
        console.log("" + testName + "\n" + (err.stack.toString()));
      }
    } else {
      console.log("\n" + this.config.summaryPass, totalPassedTests, totalTests);
    }
    return console.log('');
  };

  return ListReporter;

})(ConsoleReporter);

if (typeof module !== "undefined" && module !== null) {
  module.exports = ListReporter;
} else {
  this.joe.ListReporter = ListReporter;
}
