// es5-ext/lib/Array/prototype/index.js
'use strict';

module.exports = {
	binarySearch:   require('./binary-search'),
	clear:          require('./clear'),
	commonLeft:     require('./common-left'),
	compact:        require('./compact'),
	contains:       require('./contains'),
	copy:           require('./copy'),
	diff:           require('./diff'),
	eIndexOf:       require('./e-index-of'),
	eLastIndexOf:   require('./e-last-index-of'),
	exclusion:      require('./exclusion'),
	find:           require('./find'),
	first:          require('./first'),
	firstIndex:     require('./first-index'),
	flatten:        require('./flatten'),
	forEachRight:   require('./for-each-right'),
	group:          require('./group'),
	indexesOf:      require('./indexes-of'),
	intersection:   require('./intersection'),
	isCopy:         require('./is-copy'),
	isUniq:         require('./is-uniq'),
	last:           require('./last'),
	lastIndex:      require('./last-index'),
	remove:         require('./remove'),
	someRight:      require('./some-right'),
	uniq:           require('./uniq')
};
